### PR TITLE
Include url bar in rasterize image

### PR DESCRIPTION
**Pull request by yaakovpraisler**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [CIAC-1051](https://jira-hq.paloaltonetworks.local/browse/CIAC-1051)

## Description
Added the option to include the URL bar in the image when using the `rasterize` command.

## Screenshots
Screen Shot 2022-08-03 at 17 08 56
Screen Shot 2022-08-03 at 17 09 45

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by yaakovpraisler** [comment]
Docker PR: https://github.com/demisto/dockerfiles/pull/7665
@glicht FYI

File: Packs/rasterize/ReleaseNotes/1_2_4.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+
+#### Integrations
+##### Rasterize
+- Added an option in the command **rasterize** to include the URL bar in the screenshot.
+- Updated the Docker image to: *devdemisto/chromium:1.0.0.32875*.
```
Comment (dantavori) [inline Packs/rasterize/ReleaseNotes/1_2_4.md:5]: don't forget to update this release note too when changing to the production image

File: Packs/rasterize/Integrations/rasterize/rasterize.py [hunk 2]
```diff
@@ -133,15 +133,25 @@ def check_response(driver):
         return_err_or_warn(EMPTY_RESPONSE_ERROR_MSG)
 
 
-def init_driver(offline_mode=False):
+def init_driver_and_display(width: int, height: int, offline_mode=False, include_url=False):
     """
-    Creates headless Google Chrome Web Driver
+    Creates headless Google Chrome Web Driver and virtual display if include_url is set to True.
```
Comment (dantavori) [inline Packs/rasterize/Integrations/rasterize/rasterize.py:138]: please add arguments to docstring

File: Packs/rasterize/Integrations/rasterize/rasterize.py [hunk 3]
```diff
@@ -172,14 +182,20 @@ def find_zombie_processes():
     return zombies, ps_out
 
 
-def quit_driver_and_reap_children(driver):
+def quit_driver_and_reap_children(driver, display):
```
Comment (dantavori) [inline Packs/rasterize/Integrations/rasterize/rasterize.py:185]:
```suggestion
def quit_driver_and_display_and_reap_children(driver, display):
```

File: Packs/rasterize/Integrations/rasterize/rasterize.py [hunk 4]
```diff
@@ -248,7 +264,7 @@ def rasterize(path: str, width: int, height: int, r_type: RasterizeType = Raster
 
 
 def rasterize_webdriver(path: str, width: int, height: int, r_type: RasterizeType = RasterizeType.PNG, wait_time: int = 0,
-                        offline_mode: bool = False, max_page_load_time: int = 180, full_screen: bool = False):
+                        offline_mode: bool = False, max_page_load_time: int = 180, full_screen: bool = False, include_url: bool = False):
```
Comment (dantavori) [inline Packs/rasterize/Integrations/rasterize/rasterize.py:267]: docstring - add the include_url and a bit more information about the display

File: Packs/rasterize/Integrations/rasterize/rasterize.py [hunk 5]
```diff
@@ -275,17 +294,17 @@ def rasterize_webdriver(path: str, width: int, height: int, r_type: RasterizeTyp
...
+            output = get_image(driver, width, height, full_screen, include_url)
         return output
     finally:
-        quit_driver_and_reap_children(driver)
+        quit_driver_and_reap_children(driver, display)
 
 
 def rasterize_headless_cmd(path: str, width: int, height: int, r_type: RasterizeType = RasterizeType.PNG, wait_time: int = 0,
-                           offline_mode: bool = False, max_page_load_time: int = 180, full_screen: bool = False):
+                           offline_mode: bool = False, max_page_load_time: int = 180, full_screen: bool = False, include_url: bool = False):
```
Comment (dantavori) [inline Packs/rasterize/Integrations/rasterize/rasterize.py:307]: remove unused

**Review by dantavori — CHANGES_REQUESTED** [review]
looks good to me, @glicht can you go over this too?

**Review by glicht — COMMENTED** [review]
See my comments. Let's also add to the playbook test a test for include_url. You can use the image ocr integration to verify the result.

**Review by dantavori — APPROVED** [review]

File: Packs/rasterize/Integrations/rasterize/rasterize.py [hunk 6]
```diff
@@ -248,22 +275,29 @@ def rasterize(path: str, width: int, height: int, r_type: RasterizeType = Raster
...
     :param path: file path, or website url
     :param width: desired snapshot width in pixels
     :param height: desired snapshot height in pixels
     :param r_type: result type: .png/.pdf
     :param wait_time: time in seconds to wait before taking a screenshot
+    :param include_url: when set to True, will include the URL bar in the image result
     """
-    driver = init_driver(offline_mode)
+
+    driver, display = init_driver_and_display(width, height, offline_mode, include_url)
```
Comment (glicht) [inline Packs/rasterize/Integrations/rasterize/rasterize.py:293]: There is a scenario that display will be setup and driver will fail. This line should be moved into the try/finally and quit_driver_and_display_and_reap_children should be modified to take into account that driver may be None or in partial state.

Comment (yaakovpraisler) [reply Packs/rasterize/Integrations/rasterize/rasterize.py]: done

File: Packs/rasterize/Integrations/rasterize/rasterize.py [hunk 7]
```diff
@@ -172,15 +191,24 @@ def find_zombie_processes():
...
+    :param driver: The driver session.
+    :param display: The display session.
+
     :return: None
     """
     demisto.debug(f'Quitting driver session: {driver.session_id}')
-    driver.quit()
+
     try:
+        if driver:
```
Comment (glicht) [inline Packs/rasterize/Integrations/rasterize/rasterize.py:206]: put in try/except and log if failed

File: Packs/rasterize/Integrations/rasterize/rasterize.py [hunk 8]
```diff
@@ -172,15 +191,24 @@ def find_zombie_processes():
...
+            driver.quit()
+
+        if display:
```
Comment (glicht) [inline Packs/rasterize/Integrations/rasterize/rasterize.py:209]: put in try/except and log if failed. We want to reach the zombie process stage even if driver.quite or display.stop raise an exception

File: Packs/rasterize/Integrations/rasterize/rasterize.py [hunk 9]
```diff
@@ -26,6 +27,7 @@
 handle_proxy()
 # Make sure our python code doesn't go through a proxy when communicating with chrome webdriver
 os.environ['no_proxy'] = 'localhost,127.0.0.1'
+os.environ['HOME'] = tempfile.gettempdir()
```
Comment (glicht) [inline Packs/rasterize/Integrations/rasterize/rasterize.py:30]: Add a comment above: # Needed for cases that rasterize is running with non-root user (docker hardening)

File: Packs/rasterize/Integrations/rasterize/rasterize.py [hunk 10]
```diff
@@ -133,15 +134,50 @@ def check_response(driver):
...
+
+    Args:
+        width: desired snapshot width in pixels
+        height: desired snapshot height in pixels
+
+    Returns:
+        The  display session
+    """
+    try:
+        demisto.debug("Starting display")
```
Comment (glicht) [inline Packs/rasterize/Integrations/rasterize/rasterize.py:149]: Add the width and height in the log

File: Packs/rasterize/Integrations/rasterize/rasterize.py [hunk 11]
```diff
@@ -172,15 +208,32 @@ def find_zombie_processes():
...
+                driver.quit()
+        except Exception as edr:
+            demisto.error(f"Failed to quit driver. Error: {edr}")
+
+        try:
+            if display:
+                demisto.debug("Stopping display")
+                display.stop()
+        except Exception as edr:
+            demisto.error(f"Failed to stop display. Error: {edr}")
```
Comment (glicht) [inline Packs/rasterize/Integrations/rasterize/rasterize.py:235]: Include traceback

File: Packs/rasterize/Integrations/rasterize/rasterize.py [hunk 12]
```diff
@@ -172,15 +208,32 @@ def find_zombie_processes():
...
-    driver.quit()
+
     try:
+
+        try:
+            if driver:
+                demisto.debug(f'Quitting driver session: {driver.session_id}')
+                driver.quit()
+        except Exception as edr:
+            demisto.error(f"Failed to quit driver. Error: {edr}")
```
Comment (glicht) [inline Packs/rasterize/Integrations/rasterize/rasterize.py:228]: Include traceback

**Review by glicht — CHANGES_REQUESTED** [review]

**Review by glicht — APPROVED** [review]
Looks good. Good job.

**Review by glicht — CHANGES_REQUESTED** [review]

File: Packs/rasterize/Integrations/rasterize/rasterize.py [hunk 13]
```diff
@@ -371,6 +444,43 @@ def get_image(driver, width: int, height: int, full_screen: bool):
...
+
+        except subprocess.CalledProcessError as se:
+            demisto.error(f'Subprocess exception: {se}. Stderr: [{se.stderr}] stdout: [{se.stdout}]')
+            raise
+
+        try:
+            with open('screenshot.png', 'rb') as f:
+                image = f.read()
+        except Exception:
+            demisto.error('Failed to read the screenshot.png image.')
```
Comment (glicht) [inline Packs/rasterize/Integrations/rasterize/rasterize.py:474]: Log the exception also

**Review by glicht — APPROVED** [review]
Approved. See small comment.

File: Packs/rasterize/Integrations/rasterize/rasterize.py (hunk 13 above)
Comment (yaakovpraisler) [reply Packs/rasterize/Integrations/rasterize/rasterize.py:474]: done

**Pull request merged**